### PR TITLE
Generic QC

### PR DESCRIPTION
**Pull request by alexander-veit**
This PR adds the generic QC functionality to Tibanna. It performs basic checks on the workflow arguments and patches the QC values and metrics.zip file obtained from the pipeline to the portal.

### CODE REVIEW

File: tibanna_ffcommon/generic_qc_utils.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+from .exceptions import (
+    GenericQcException
+)
+
+def check_qc_workflow_args(input_file_args, generic_qc_args):
+    """
+    This function performs basic sanity checks on the QC and input files
+    """
```
Comment (willronchetti) [inline tibanna_ffcommon/generic_qc_utils.py:8]: For new code I prefer type annotations and more info in the docstring on structure etc - PyCharm will generate these for you quickly with minimal overhead. I won't comment this everywhere but it applies in several places.

File: tibanna_ffcommon/portal_utils.py [hunk 2]
```diff
@@ -74,6 +87,14 @@
...
 logger.debug(pkg_resources.get_distribution('dcicutils').version)
 
+OUTPUT_ARG_TYPE_LIST = [OUTPUT_PROCESSED_FILE,
+                        OUTPUT_REPORT_FILE,
+                        OUTPUT_QC_FILE,
+                        GENERIC_QC_FILE,
+                        OUTPUT_TO_BE_EXTRA_INPUT_FILE]
+
+# File types that are uploaded to the portal (and S3)
+PROCESSED_FILE_TYPES = [OUTPUT_PROCESSED_FILE, GENERIC_QC_FILE]
```
Comment (willronchetti) [inline tibanna_ffcommon/portal_utils.py:97]: I noticed this looking at the portal - is the implication here that there are two metadata records for QC files? One processed file and one QC object?

File: tibanna_ffcommon/portal_utils.py [hunk 3]
```diff
@@ -1220,10 +1242,80 @@ def update_input_extras(self):
...
+            qc_arg_zipped_s3_url = f"https://{self.outbucket}.s3.amazonaws.com/{self.file_key(qc_arg_zipped['workflow_argument_name'])}" if qc_arg_zipped else None
+
+            # The folling will create a new QualityMetricGeneric item in the portal
+            qmg_uuid = str(uuid4())
+            try:
+                post_dict = {
+                    'uuid': qmg_uuid,
+                    'institution': self.ff_meta.institution,
+                    'project': self.ff_meta.project,
+                }
```
Comment (willronchetti) [inline tibanna_ffcommon/portal_utils.py:1291]: This is CGAP specific (project/institution) no? So does this really belong in tibanna_ffcommon? How would this work in SMaHT?

**Review by willronchetti — COMMENTED** [review]
I think the comment on hard coding project/institution merits addressing seeing as we will need this functionality for SMaHT

File: tibanna_ffcommon/generic_qc_utils.py (hunk 1 above)
Comment (alexander-veit) [reply tibanna_ffcommon/generic_qc_utils.py]: I added type hints and more detailed docstrings to all new functions

File: tibanna_ffcommon/portal_utils.py (hunk 2 above)
Comment (alexander-veit) [reply tibanna_ffcommon/portal_utils.py:97]: Yes, there is one QC file (processed file) and the QualityMetricsGeneric object that contains the data of that QC file and a link to the metrics report zip file.

File: tibanna_ffcommon/portal_utils.py [hunk 4]
```diff
@@ -967,6 +984,11 @@ def s3(self, argname):
     def read(self, argname):
         """This function is useful for reading md5 report of qc report"""
         return self.s3(argname).read_s3(self.file_key(argname)).decode('utf-8', 'backslashreplace')
+    
+    def read_json_from_s3(self, argname):
+        """This function is useful for reading generic QC files"""
+        file_content = self.s3(argname).read_s3(self.file_key(argname)).decode('utf-8')
+        return json.loads(file_content)
```
[the remark this answers is not in the document]
Comment (alexander-veit) [reply tibanna_ffcommon/portal_utils.py]: Added a `try` `except`.

File: tibanna_ffcommon/portal_utils.py (hunk 3 above)
Comment (alexander-veit) [reply tibanna_ffcommon/portal_utils.py]: This has been generalized.

**Review by willronchetti — APPROVED** [review]

**Pull request merged**